File: code/design-patterns/structural/flyweight/python/flyweight.py
from typing import List


class TreeType:
    def __init__(self, name, color, texture):
        self.name = name
        self.color = color
        self.texture = texture


class Tree:
    def __init__(self, x, y, tree_type: TreeType):
        self.x = x
        self.y = y
        self.tree_type = tree_type

    def draw(self, canvas, x, y):
        print("Drawing on canvas")


class TreeFactory:
    tree_types: List[TreeType] = []

    def get_tree_type(self, name, color, texture) -> TreeType:
        # Look for existing tree type
        for tree_type in self.tree_types:
            if (
                tree_type.name == name
                and tree_type.color == color
                and tree_type.texture == texture
            ):
                return tree_type
        # If not found, create a new one and add it to the list
        tree_type = TreeType(name, color, texture)
        self.tree_types.append(tree_type)
        return tree_type


class Forest:

    def __init__(self):
        self.trees: List[Tree] = []

    def plant_tree(self, x, y, name, color, texture) -> Tree:
        tree_type = TreeFactory().get_tree_type(name, color, texture)
        tree = Tree(x, y, tree_type)
        self.trees.append(tree)
        return tree

    def draw(self, canvas):
        for tree in self.trees:
            tree.draw(canvas, tree.x, tree.y)

File: code/design-patterns/structural/flyweight/go/flyweight.go
package flyweight

import (
	"fmt"
)

// TreeType represents the shared intrinsic state of a tree
type TreeType struct {
	Name    string
	Color   string
	Texture string
}

// Tree represents a tree object with extrinsic state (x, y) and intrinsic state (TreeType)
type Tree struct {
	X        int
	Y        int
	TreeType *TreeType
}

// Draw simulates drawing the tree on a canvas
func (t *Tree) Draw(canvas string) {
	fmt.Printf("Drawing tree on %s at (%d, %d) with type (%s, %s, %s)\n",
		canvas, t.X, t.Y, t.TreeType.Name, t.TreeType.Color, t.TreeType.Texture)
}

// TreeFactory manages the creation and reuse of TreeType instances
type TreeFactory struct {
	treeTypes []*TreeType
}

// GetTreeType returns an existing TreeType if found, or creates a new one
func (f *TreeFactory) GetTreeType(name, color, texture string) *TreeType {
	for _, treeType := range f.treeTypes {
		if treeType.Name == name && treeType.Color == color && treeType.Texture == texture {
			return treeType
		}
	}
	treeType := &TreeType{Name: name, Color: color, Texture: texture}
	f.treeTypes = append(f.treeTypes, treeType)
	return treeType
}

// Forest represents a collection of trees
type Forest struct {
	trees       []*Tree
	treeFactory *TreeFactory
}

// PlantTree plants a new tree in the forest
func (f *Forest) PlantTree(x, y int, name, color, texture string) {
	treeType := f.treeFactory.GetTreeType(name, color, texture)
	tree := &Tree{X: x, Y: y, TreeType: treeType}
	f.trees = append(f.trees, tree)
}

// Draw draws all the trees in the forest
func (f *Forest) Draw(canvas string) {
	for _, tree := range f.trees {
		tree.Draw(canvas)
	}
}
